### PR TITLE
ALL: Remove all strcat, strcpy, sprintf and vsprintf uses and forbid their usage

### PR DESCRIPTION
**Pull request by lephilousophe**
These functions are completely unsafe and should not be used anymore.

To ease the transition a 2 arguments version of these functions is introduced which takes an array reference.
This allows the compiler to infer the size argument of the original function and avoids errors during the conversion.

For now, common code, backend code (Linux) and every engines are converted.
As I can't test everything, a thorough review will be needed.

### CODE REVIEW

File: engines/dm/menus.cpp [hunk 1]
```diff
@@ -1628,8 +1628,11 @@ void MenuMan::printMessageAfterReplacements(const char *str) {
 				replacementString = championMan._champions[_vm->ordinalToIndex(championMan._actingChampionOrdinal)]._name;
 
 			*curCharacter = '\0';
-			strcat(outputString, replacementString);
-			curCharacter += strlen(replacementString);
+			size_t ln = Common::strlcat(outputString, replacementString);
+			if (ln >= sizeof(outputString)) {
```
Comment (sev-) [inline engines/dm/menus.cpp:1632]: Interesting solution, what is going on here? What is the side effect of ending execution here?

**Review by sev- — COMMENTED** [review]
Looked through code till GLK. Only two minor notes/questions.

File: engines/agos/string_pn.cpp [hunk 2]
```diff
@@ -117,7 +119,7 @@ void AGOSEngine_PN::pcl(const char *s) {
 	if (strchr(s, '\n') == nullptr) {
 		for (char *str = _sb; *str; str++)
 			windowPutChar(_windowArray[_curWindow], *str);
-		strcpy(_sb, "");
+		_sb[0] = '\0'; /* Reset to empty string */
```
[the remark this answers is not in the document]
Comment (lephilousophe) [reply engines/agos/string_pn.cpp]: Done.

File: engines/dm/menus.cpp (hunk 1 above)
Comment (lephilousophe) [reply engines/dm/menus.cpp:1632]: Indeed, I made it an error. It's cleaner.

File: engines/ags/engine/ac/string.cpp [hunk 3]
```diff
@@ -325,11 +326,7 @@ buffr[199]=0;
 void my_strncpy(char *dest, const char *src, int len) {
 	// the normal strncpy pads out the string with zeros up to the
 	// max length -- we don't want that
-	if (strlen(src) >= (unsigned)len) {
-		strncpy(dest, src, len);
-		dest[len] = 0;
-	} else
-		strcpy(dest, src);
+	Common::strlcpy(dest, src, len);
```
Comment (criezy) [inline engines/ags/engine/ac/string.cpp:329]:
```c++
	// This function takes the length of the buffer minus 1 (i.e. the length does not include the null terminator).
	Common::strlcpy(dest, src, len + 1);
````

You can see it in the old implementation `dest[len] = 0;` as well as in places where it is called (e.g. `my_strncpy(destt, text, _G(MAXSTRLEN) - 1);`).

Comment (lephilousophe) [reply engines/ags/engine/ac/string.cpp:329]: Thanks! I will fix it when I have time to continue the PR.

Comment (lephilousophe) [reply engines/ags/engine/ac/string.cpp]: Done.

**Comment by lephilousophe** [comment]
This PR should be complete except backends code.
I will let Github builds spot problems in Mac and Windows builds.
Next week-end I will try to make a run on my own buildbot.

**Comment by lephilousophe** [comment]
This is a new version of the same thing but instead of using strlcpy/strlcat and silently cropping fields, I created new functions strcpy_s/strcat_s, loosely inspired by official C11 functions, which issue a warning if there is an overflow and if arguments have invalid values.
I now plan to tackle sprintf usage.

**Comment by sev-** [comment]
Time to merge before it got bitrot?

**Review by sev- — APPROVED** [review]

**Comment by criezy** [comment]
I like the idea for this change.
Given the amount of changes it's difficult to review everything, so maybe we could indeed merge it and rely on compilers and testing to catch potential bugs that might have crept in.

**Comment by lephilousophe** [comment]
I got busy with Libretro shaders but this should settle down and I could continue with `sprintf`.

**Comment by lephilousophe** [comment]
I just finished to migrate from sprintf and vsprintf too.
Let's wait for Github actions to build and I will try to start a build on my buildbot too.
I hope no bugs have been added in the process.

**Comment by lephilousophe** [comment]
This now builds correctly on my buildbot

**Comment by lotharsm** [comment]
This also builds and runs fine on my MSYS2/Mingw64 toolchain.

**Comment by sev-** [comment]
BIG GREEN BUTTON PUSH.

**Pull request merged**